### PR TITLE
gh-70039: smtplib: store the server name in ._host in .connect()

### PR DESCRIPTION
**Pull request by nmartensen**
Original patch by gigaplastik, extended with a few more tests.

Addresses:

* Issue: gh-70039

* Issue: gh-80275

### CODE REVIEW

**Comment by nmartensen** [comment]
The PR adds 6 tests in addition to the functional change. Most of them fail without the change:
- In Lib/test/test_smtplib.py:DebuggingServerTests
  - test_host_port_host  [fails] *
  - test_explicit_connect  [fails]
  - text_explicit_connect_host_port  [fails]
- In Lib/test/test_smtpnet.py:SmtpTest
  - test_connect_host_port_starttls
  - test_explicit_connect_starttls [fails]
  - test_explicit_connect_host_port_starttls [fails]
 
The one that does not fail describes a situation that could sometimes also fail in practice before the change, see e.g. <https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=926900>. The test marked with '*' above is meant to ensure this failure case won't be accidentally reintroduced in potential future changes.

**Comment by jahs** [comment]
Hi all, I'd like to echo the above as this is still an issue (going back to 2015). Caused an issue this week (December 2025) due to AWS SES eu-west-2 applying strict parsing in TLS of the self._host string (which they rolled back, in part due to this client bug). Specifically, the SMTP("host:port") constructor was used and it failed to connect, as the connect() parsing currently does not update self._host, which is then used in starttls().

File: Lib/test/test_smtplib.py [hunk 1]
```diff
@@ -282,6 +282,34 @@ def testBasic(self):
...
+        self.assertEqual(smtp._host, HOST)
+        smtp.quit()
+
+    def test_explicit_connect_host_port(self):
+        # create instance without arguments
+        smtp = smtplib.SMTP(local_hostname='localhost',
+                            timeout=support.LOOPBACK_TIMEOUT)
+        # explicit .connect with host:port notation must set ._host, too
+        smtp.connect(f"{HOST}:{self.port}")
+        self.assertEqual(smtp._host, HOST)
```
Comment (bitdancer) [inline Lib/test/test_smtplib.py]: I think these tests should be dropped.  At this time, `_host` is an implementation detail, and I gather they were originally implemented because of the lack of network tests.  The bug is starttls failing when connect is used explicitly, and the smtpnet tests check that.

File: Misc/NEWS.d/next/Library/2024-02-10-21-25-22.gh-issue-70039.6wvcAP.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+In smtplib.SMTP, set ._host in .connect() and not immediately during instantiation. This allows .starttls() to work if the instance was created without arguments and host and port were only provided via an explicit .connect() call. If will also set ._host to the correct host value if the host and port were specified using the 'host:port' notation, which according to the documentation is supported by .connect().
```
Comment (bitdancer) [inline Misc/NEWS.d/next/Library/2024-02-10-21-25-22.gh-issue-70039.6wvcAP.rst]: I would rephrase this to not talk about the implementation detail of `_host`: "Fixed bug where :meth:`SMTP.starttls` could fail if :meth:`SMTP.connection` is called explicitly rather than implicitly."  Everything else is implementation details and belongs in the commit message.

**Review by bitdancer — CHANGES_REQUESTED** [review]

**Comment by nmartensen** [comment]
I have made the requested changes; please review again.

**Comment by nmartensen** [comment]
There seem to be failing tests with this PR now. I have looked into them but fail to see what is needed to fix them. Please let me know if you know more.

**Comment by bitdancer** [comment]
I *think* all you need to do is update to the tip of main.  Let's try that, at least.

**Review by bitdancer — APPROVED** [review]
